### DISCUSSION TITLE
conlist in Link type

### DISCUSSION THREAD

**Discussion by opaniagu** [BeanieODM/beanie#549] [Feature Request]
I need this pattern:

my_list: conlist(Optional[List[Link[Principal]]], unique_items=True)

I get an error:

value is not a valid list (type=type_error.list)

Thanks

**Comment by roman-right** [comment]
Hey! I didn't try `conlist` with links yet. Can you please share a one-module sample with me be able to run it and see, what I can do to support your case?

**Comment by opaniagu** [comment]
```code
class User(Document):
    username: Indexed(str, unique=True)  # type: ignore
    email: Optional[str]
    principals: Optional[List[Link[Principal]]]

    class Settings:
        use_revision = False

    async def add_principal(self, principal: Principal):
        if principal not in self.principals:
            self.principals.append(principal)
            await self.save(link_rule=WriteRules.WRITE)

class Principal(Document):
    identifier: Identifier = Identifier.Role
    name: Indexed(str, unique=True)  # type: ignore
    description: str = ""
```
The field 'principals' is like a User 'Role'. 
One user can have many roles, for this reason I define a List w/ 'conslit', but I want that no repeat (conlist using unique_items)

I solved, coding a setter add_principal()
